File: wrappers/go/garry-go/encoder.go
package garry

import "errors"

var (
	ErrKeyPartTooLong = errors.New("garry: key part exceeds 255 bytes")
	ErrKeyTruncated   = errors.New("garry: key data truncated")
)

func EncodeKey(parts ...string) []byte {
	total := 0
	for _, p := range parts {
		b := []byte(p)
		if len(b) > 255 {
			panic(ErrKeyPartTooLong)
		}
		total += 1 + len(b)
	}
	out := make([]byte, 0, total)
	for _, p := range parts {
		b := []byte(p)
		out = append(out, byte(len(b)))
		out = append(out, b...)
	}
	return out
}

func DecodeKey(key []byte) []string {
	var parts []string
	off := 0
	for off < len(key) {
		if off+1 > len(key) {
			panic(ErrKeyTruncated)
		}
		n := int(key[off])
		off++
		if off+n > len(key) {
			panic(ErrKeyTruncated)
		}
		parts = append(parts, string(key[off:off+n]))
		off += n
	}
	return parts
}

func CombineKey(prefix []byte, additionalParts ...string) []byte {
	existing := DecodeKey(prefix)
	all := make([]string, 0, len(existing)+len(additionalParts))
	all = append(all, existing...)
	all = append(all, additionalParts...)
	return EncodeKey(all...)
}

File: wrappers/js/garry/lib/encoder.js
function encodeKey(...parts) {
    const bufs = parts.map(p => {
        if (typeof p === 'string') return Buffer.from(p, 'utf8');
        if (Buffer.isBuffer(p)) return p;
        throw new TypeError('Key part must be string or Buffer');
    });
    const lens = bufs.map(b => {
        if (b.length > 255) throw new RangeError(`Key part exceeds 255 bytes (${b.length})`);
        return b.length;
    });
    const totalLen = lens.reduce((s, l) => s + 1 + l, 0);
    const out = Buffer.alloc(totalLen);
    let off = 0;
    for (let i = 0; i < bufs.length; i++) {
        out[off++] = lens[i];
        bufs[i].copy(out, off);
        off += lens[i];
    }
    return out;
}

function decodeKey(key) {
    const buf = Buffer.isBuffer(key) ? key : Buffer.from(key, 'utf8');
    const parts = [];
    let off = 0;
    while (off < buf.length) {
        const len = buf[off++];
        if (off + len > buf.length) throw new Error('Key data truncated');
        parts.push(buf.slice(off, off + len).toString('utf8'));
        off += len;
    }
    return parts;
}

function combineKey(prefix, ...additionalParts) {
    const prefixParts = decodeKey(prefix);
    return encodeKey(...prefixParts, ...additionalParts);
}

module.exports = { encodeKey, decodeKey, combineKey };
